source/GameAction.cpp

```diff
@@ -0,0 +1,228 @@
+}
+
+
+
+// Load a single child at a time, used for streamlining MissionAction::Load.
+void GameAction::LoadSingle(const DataNode &child, const string &missionName)
+{
+	const string &key = child.Token(0);
+	bool hasValue = (child.Size() >= 2);
+	
+	if(key == "log")
+	{
+		bool isSpecial = (child.Size() >= 3);
+		string &text = (isSpecial ?
+			specialLogText[child.Token(1)][child.Token(2)] : logText);
+		Dialog::ParseTextNode(child, isSpecial ? 3 : 1, text);
+	}
+	else if(key == "payment")
+	{
+		if(child.Size() == 1)
+			paymentMultiplier += 150;
+		if(child.Size() >= 2)
+			payment += child.Value(1);
+		if(child.Size() >= 3)
+			paymentMultiplier += child.Value(2);
+	}
+	else if(key == "event" && hasValue)
+	{
+		int minDays = (child.Size() >= 3 ? child.Value(2) : 0);
+		int maxDays = (child.Size() >= 4 ? child.Value(3) : minDays);
+		if(maxDays < minDays)
+			swap(minDays, maxDays);
+		events[GameData::Events().Get(child.Token(1))] = make_pair(minDays, maxDays);
+	}
+	else if(key == "fail")
+	{
+		string toFail = child.Size() >= 2 ? child.Token(1) : missionName;
+		fail.insert(toFail);
+		// Create a GameData reference to this mission name.
+		GameData::Missions().Get(toFail);
+	}
+	else
+		conditions.Add(child);
+}
+
+
+
+void GameAction::Save(DataWriter &out) const
+{
+	if(!logText.empty())
+	{
+		out.Write("log");
+		out.BeginChild();
+		{
+			// Break the text up into paragraphs.
+			for(const string &line : Format::Split(logText, "\n\t"))
+				out.Write(line);
+		}
+		out.EndChild();
+	}
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+		{
+			out.Write("log", it.first, eit.first);
+			out.BeginChild();
+			{
+				// Break the text up into paragraphs.
+				for(const string &line : Format::Split(eit.second, "\n\t"))
+					out.Write(line);
+			}
+			out.EndChild();
+		}
+	if(payment)
+		out.Write("payment", payment);
+	for(const auto &it : events)
+	{
+		if(it.second.first == it.second.second)
+			out.Write("event", it.first->Name(), it.second.first);
+		else
+			out.Write("event", it.first->Name(), it.second.first, it.second.second);
+	}
+	for(const auto &name : fail)
+		out.Write("fail", name);
+	
+	conditions.Save(out);
+}
+
+
+
+bool GameAction::IsEmpty() const
+{
+	return empty;
+}
+
+
+
+int GameAction::Payment() const
+{
+	return payment;
+}
+
+
+
+// Do the actions of the GameAction. If this GameAction is a part of a conversations,
+// then some behavior may be different than from a MissionAction.
+void GameAction::Do(PlayerInfo &player, bool conversation) const
+{
+	if(!logText.empty())
+		player.AddLogEntry(logText);
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+			player.AddSpecialLog(it.first, eit.first, eit.second);
+	
+	if(payment)
+	{
+		// Conversation actions don't block a mission from offering if a
+		// negative payment would drop the player's account balance below
+		// zero, so payment is handled differently within conversations.
+		if(conversation)
```

Comment:
You could use the Conversations code both for the conversations and the non-conversations case (since you are sure in the non-conversations case that `account + payment >= 0` is always true).